pydis_site/apps/home/signals.py

```diff
@@ -0,0 +1,192 @@
+            return
+
+        try:
+            user: DiscordUser = DiscordUser.objects.get(id=int(account.uid))
+        except DiscordUser.DoesNotExist:
+            return
+
+        self._apply_groups(user, account)
+
+    def social_account_removed(self, sender: Type[SocialLogin], **kwargs) -> None:
+        """Handle signals relating to removal of social accounts."""
+        account: SocialAccount = kwargs["socialaccount"]
+        provider: Provider = account.get_provider()
+
+        if not isinstance(provider, DiscordProvider):
+            return
+
+        try:
+            user: DiscordUser = DiscordUser.objects.get(id=int(account.uid))
+        except DiscordUser.DoesNotExist:
+            return
+
+        self._apply_groups(user, account, deletion=True)
+
+    def mapping_model_deleted(self, sender: Type[RoleMapping], **kwargs) -> None:
+        """Handle signals related to the deletion of Role Mapping model entries."""
+        instance: RoleMapping = kwargs["instance"]
+
+        for user in instance.group.user_set.all():
+            user.groups.remove(instance.group)
+
+    def mapping_model_updated(self, sender: Type[RoleMapping], **kwargs) -> None:
+        """Handle signals related to the updating of Role Mapping model entries."""
+        instance: RoleMapping = kwargs["instance"]
+        raw: bool = kwargs["raw"]
+
+        if raw:
+            # Fixtures are being loaded, so don't touch anything
+            return
+
+        old_instance: Optional[RoleMapping] = None
+
+        if instance.id is not None:
+            # We don't try to catch DoesNotExist here because we can't test for it,
+            # it should never happen (unless we have a bad DB failure)
+
+            old_instance = RoleMapping.objects.get(id=instance.id)
+
+        if old_instance:
+            self.mapping_model_deleted(RoleMapping, instance=old_instance)
+
+        accounts = SocialAccount.objects.filter(
+            uid__in=(u.id for u in instance.role.user_set.all())
+        )
+
+        for account in accounts:
+            account.user.groups.add(instance.group)
+
+    def user_model_updated(self, sender: Type[DiscordUser], **kwargs) -> None:
+        """Handle signals related to the updating of Discord User model entries."""
+        instance: DiscordUser = kwargs["instance"]
+        raw: bool = kwargs["raw"]
+
+        # `update_fields` could be used for checking changes, but it's None here due to how the
+        # model is saved without using that argument - so we can't use it.
+
+        if raw:
+            # Fixtures are being loaded, so don't touch anything
+            return
+
+        try:
+            account: SocialAccount = SocialAccount.objects.get(
+                uid=str(instance.id), provider=DiscordProvider.id
+            )
+        except SocialAccount.DoesNotExist:
+            return  # User has never logged in with Discord on the site
+
+        self._apply_groups(instance, account)
+
+    def _apply_groups(
```

Comment:
Don't we lint for docstrings on functions like this?

Either way I think there should be a docstring.

Reply:
We don't lint for docstrings on private methods - just public ones - but I'm happy to add one.